File: Chinpokomon/ChinpokomonPython/chinpokomon/chinpokomon.py
from logger.logger import Logger
import random


class Chinpokomon():

    def __init__(self, name, health, nature, attacks=[]):
        self._logger = Logger()
        self._name = name
        self._health = health
        self._nature = nature
        self._max_health = health
        self._attacks = attacks

        for attack in attacks:
            attack.chinpo = self

    @property
    def name(self):
        return self._name

    @property
    def nature(self):
        return self._nature

    @property
    def health(self):
        return self._health

    @health.setter
    def health(self, health):
        self._health = health

    @property
    def max_health(self):
        return self._max_health

    @property
    def attacks(self):
        return self._attacks

    @property
    def logger(self):
        return self._logger

    def lear_attack(self, attack):
        attack.chinpo = self
        self._attacks.append(attack)

    def lear_attacks(self, attacks):
        for attack in attacks:
            self.lear_attack(attack)

    def attack(self, chinpo):
        random_attack = self.random_attack()
        random_attack.activate(chinpo)

    def recieve_damage(self, damage):
        if ((self.health - damage) < 0):
            self.health = 0
        else:
            self.health = self.health - damage
        self.logger.info(self.name + " recibio " +
                         str(damage) + " puntos de daño.")
        self.logger.info("Puntos de vida de " +
                         str(self.name) + " " + str(self.health))

    def recieve_healing(self, healing):
        if ((self.health + healing) > self.max_health):
            self.health = self.max_health
        else:
            self.health = self.health + healing
        self.logger.info(self.name + " recibio " +
                         str(healing) + " puntos de curación.")
        self.logger.info("Puntos de vida de " +
                         self.name + " " + str(self.health))

    def random_attack(self):
        return random.choice(self.attacks)

    def restore_health(self):
        self.health = self.max_health

    def ko(self):
        return self.health == 0

    def has_advantage_against(self, chinpo):
        return self.nature.has_advantage_against(chinpo.nature)

    def has_disadvantage_against(self, chinpo):
        return self.nature.has_disadvantage_against(chinpo.nature)

File: Chinpokomon/ChinpokomonJava/src/main/java/poo3/chinpokomon/Chinpokomon.java
package poo3.chinpokomon;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import poo3.attack.Attack;
import poo3.chinpokomon.nature.NatureType;
import poo3.logger.Logger;

public abstract class Chinpokomon {
	protected Integer health;
	protected Integer maxHealth;
	protected List<Attack> attacks;
	protected String name;
	protected NatureType nature;
	protected Logger log;
	protected Random random;
	
	public Chinpokomon(String name, Integer health, NatureType nature){
		this.name = name;
		this.nature = nature;
		this.health = health;
		this.maxHealth = health;
		this.log = Logger.getInstance();
		this.random = new Random();
		this.attacks = new ArrayList<Attack>();
	}
	
	public Chinpokomon(String name, Integer health, NatureType nature, List<Attack> attacks) {
		this(name, health, nature);
		this.attacks.addAll(attacks);
	}
	
	public void learnAttack(Attack attack) {
		this.attacks.add(attack);
	}
	
	public void attack(Chinpokomon chinpo) {
		Attack randomAttack = this.randomAttack();
		randomAttack.attack(this, chinpo);
	}
	
	public void recieveDamage(Integer damage) {
		if((this.health - damage) < 0) {
			this.health = 0;
		} else {
			this.health -= damage;
		}
		log.info(this.name + " recibio " + damage + " puntos de daño.");
	}
	
	public void recieveHealing(Integer healing) {
		if((this.health + healing) > this.maxHealth) {
			this.health = this.maxHealth;
		} else {
			this.health += healing;
		}
		log.info(this.name + " recibio " + healing + " puntos de curación.");
	}
	
	protected Attack randomAttack() {
		return this.attacks.get(random.nextInt(this.attacks.size()));
	}
	
	public void restoreHealth() {
		this.health = this.maxHealth;
	}
	
	public Boolean KO() {
		return this.health == 0;
	}
	
	public Integer getHealth() {
		return this.health;
	}
	
	public String getName() {
		return this.name;
	}
	
	public NatureType getNature() {
		return this.nature;
	}
	
	public Boolean hasAdvantageAgainst(Chinpokomon chinpo) {
		return this.nature.hasAdvantageAgainst(chinpo.getNature());
	}
	
	public Boolean hasDisadvantageAgainst(Chinpokomon chinpo) {
		return this.nature.hasDisadvantageAgainst(chinpo.getNature());
	}
}
